### DISCUSSION TITLE
Community toolkit button within a listbox doesn't work (RelayCommand)

### DISCUSSION THREAD

**Discussion by Yveske1966** [AvaloniaUI/Avalonia#14071] [Q&A] [answered]
Constants.cs:
`    public static ViewModelBase CurrentViewModel { get; set; }`

ViewModelBase.cs:
`    public ViewModelBase () {
        Constants.CurrentViewModel = this;
    }`

DownloadsView.axaml:
![image](https://github.com/AvaloniaUI/Avalonia/assets/68047793/5ab9ae71-dbd1-4f16-bd24-ae616452330a)

DownloadsView.axaml.cs:
`    public DownloadsView () {
        DataContext = (DownloadsViewModel) Constants.CurrentViewModel;
        InitializeComponent ();
    }`

DownloadsViewModel.cs:
`    [RelayCommand]
    private async Task DownloadStatusClickedAsync () {
        return;
    }`

- In Maui I could use the OnClick (which no longer exists in Avalonia) of the button and the code behind to access the ViewModel.
- The DataContext was empty when entering the constructor of DownloadsView, so I'm setting it manually, using the static filled in by the ViewModelBase.
- I'm using $parent[UserControl] to try to persuade Avalonia to look in the correct position for the RelayCommand to no avail, the button seems to remain not enabled.

![image](https://github.com/AvaloniaUI/Avalonia/assets/68047793/0be8ea34-ce0d-4ca4-baf7-6d38720818c2)

What am I doing wrong here ?

**Comment by timunie** [comment]
You can try to bind to named element instead of parent. Syntax is #root.DataContext.... where root is the name of the UserControl

**Reply by Yveske1966** [reply]
Added x:Name="root" to the UserControl and replace $parent[UserControl] with #root, but got 'Internal compiler error while transforming node XamlX.Ast.XamlAstObjectNode: Avalonia.Data.Core.ExpressionParseException: Element name expected after '#'

**Reply by timunie** [reply]
Please send the new Binding code here.

**Reply by Yveske1966** [reply]
Command="{Binding #root.((vm:DownloadsViewModel) DataContext).DownloadStatusClickedCommand}"
also tried
Command="{Binding #root.DataContext.DownloadStatusClickedCommand}"
which results in Unable to resolve property or method of name 'DownloadStatusClickedCommand' on type 'System.Object'.

**Reply by timunie** [reply, marked as answer]
Could it be there's a whitespace which confuses the parser? `...) DataContext` the one behind the bracket

**Reply by Yveske1966** [reply]
Yep ! It doesn't like whitespace !

Thanks a lot and Happy New Year 2024 !

**Reply by thevortexcloud** [reply]
I have run into this problem before. It really does feel like something that should work.

**Reply by Yveske1966** [reply]
The real problem to me is that ListBox.ItemSource hides the ViewModel and there is no easy way to circumvent it.
It would have been far more easier to have to reference fields using table.field while leaving the ViewModel alone for commands.
Maui is better in that regard, but is not production ready (yet).

**Reply by thevortexcloud** [reply]
The automatic data context switching can be useful, but yeah it can be annoying too. But this is also how WPF behaves. Which is also probably why Avalonia does it that way. Having some sort of shortcut like `$parent` that gave access to the control's datacontext would be very handy.

**Reply by timunie** [reply]
> The real problem to me is that ListBox.ItemSource hides the ViewModel and there is no easy way to circumvent it.
> It would have been far more easier to have to reference fields using table.field while leaving the ViewModel alone for commands.
> Maui is better in that regard, but is not production ready (yet).

It doesn't hide the ViewModel, it does what is expected. The ViewModel of the ItemsControl Items should be the ItemViewModel, not the MainViewModel. I don't get how MAUI could do "better" here?

**Reply by Yveske1966** [reply]
In the code on top I can simple write Command={Binding GoToPreviousPageCommand}
ListBox binds its ItemSource to an ObservableCollection called Downloads in the ViewModel.
After that I have to reference the UserControl to get to the DataContext.
On top of that I need to manually fill in that DataContext via a static (Multi-threading ????).

On Maui on the other hand the command:
await Shell.Current.GoToAsync (nameof (DownloadsPage), new Dictionary<string, object> () {{ "MyQuery", myQuery }});
Jumps to the Downloads page and hands me the DownloadsViewModel.
It also (optionally) allows me to pass extra parameters (as displayed above). -> still need to tackle that one later !
Using Button.OnClicked I can access the correct ViewModel function in code-behind. Easy Peasy !
